libmattermost.c

```diff
@@ -469,12 +469,195 @@ int mkd_line(char *, int, char **, int);
+	{
+	.find = "<font size=\"7\">(.*)</font>",
+	.repl = " # \\1",
+	.regex = NULL,
+	},
+	// title2 
+	{
+	.find = "<font size=\"6\">(.*)</font>",
+	.repl = " ## \\1",
+	.regex = NULL,
+	},
+	// title3
+	{
+	.find = "<font size=\"5\">(.*)</font>",
+	.repl = " ### \\1",
+	.regex = NULL,
+	},
+	// title4 
+	{
+	.find = "<font size=\"4\">(.*)</font>",
+	.repl = " #### \\1",
+	.regex = NULL,
+	},
+	// horizontal line
+	{
+	.find = "<hr>",
+	.repl = "\n---\n",
+	.regex = NULL,
+	},
+	// blockquote
+	{
+	.find = "^ *&gt;(.*)$",
+	.repl = ">\\1",
+	.regex = NULL,
+	},
+};
+
+#define MM_MAX_REGEX 9
+
+static MattermostRegexElement mm_regexes[MM_MAX_REGEX]={
+	// line break 
+	{
+	.find = "<br>",
+	.repl = "\n",
+	.regex = NULL,
+	},
+	// (inline) code block 
+	{
+	.find = "<code>(.*)</code>",
+	.repl = "<font back=\"#E1E1E1\">\\1</font>",
+	.regex = NULL,
+	},
+	// title1
+	{
+	.find = "^ *# +(.*)($|<br>)",
+	.repl = "<font size=\"7\"><b>\\1</b></font>",
+	.regex = NULL,
+	},
+	// title2
+	{
+	.find = "^ *## +(.*)$",
+	.repl = "<font size=\"6\"><b>\\1</b></font>",
+	.regex = NULL,
+	},
+	// title3
+	{
+	.find = "^ *### +(.*)$",
+	.repl = "<font size=\"5\"><b>\\1</b></font>",
+	.regex = NULL,
+	},
+	// title4-6 (normal font size is 3)
+	{
+	.find = "^ *#####?#? +(.*)$",
+	.repl = "<font size=\"4\"<b>\\1</b></font>",
+	.regex = NULL,
+	},
+	// horizontal line
+	{	
+	.find = "^ *-{3,}$",
+	.repl = "<hr>",
+	.regex = NULL,
+	},
+	// blockquote
+	{
+	.find = "^ *(&gt;|>)(.*)$",
+	.repl = "<font size=\"6\"><b>\"</b></font>\\2", //0x93 ?
+	.regex = NULL,
+	},
+	// strikethrough
+	{
+	.find = "<del>(.*)</del>",
+	.repl = "<s>\\1</s>",
+	.regex = NULL,
+	},
+};
+
+static void 
+mm_purple_xhtml_im_html_init(void)
+{
+	gint i;
+
+	for (i=0;i< MM_MAX_REGEX; i++) {
+		mm_regexes[i].regex = g_regex_new(mm_regexes[i].find, G_REGEX_CASELESS|G_REGEX_DOTALL|G_REGEX_OPTIMIZE|G_REGEX_MULTILINE|G_REGEX_UNGREEDY, G_REGEX_MATCH_NOTEMPTY, NULL);
+	}
+	for (i=0;i< MM_MAX_REV_REGEX; i++) {
+		mm_rev_regexes[i].regex = g_regex_new(mm_rev_regexes[i].find, G_REGEX_CASELESS|G_REGEX_DOTALL|G_REGEX_OPTIMIZE|G_REGEX_MULTILINE|G_REGEX_UNGREEDY, G_REGEX_MATCH_NOTEMPTY, NULL);
+	}
+
+}
+
+
+static gchar *
+mm_purple_html_to_xhtml_im_parse(MattermostAccount *ma, const gchar *html)
+{
+	gint i;
+	gchar *input = NULL;
+	gchar *output = NULL;
+
+	if(!purple_account_get_bool(ma->account, "use-markdown", TRUE)) {
+		return g_strdup(html);
+	}
+ 
+	if (html == NULL) {
+		return NULL;
+	}
+
+	input = g_strdup(html);
+	for (i=0;i< MM_MAX_REGEX; i++) {
+		output = g_regex_replace(mm_regexes[i].regex, input, -1, 0, mm_regexes[i].repl, G_REGEX_MATCH_NOTEMPTY, NULL);
+		g_free(input);
+		input = g_strdup(output);
+		g_free(output);
+	}
+	
+	return g_strdup(input);
+}
+
+static gchar *
+mm_purple_xhtml_im_to_html_parse(MattermostAccount *ma, const gchar *xhtml_im)
+{
+	gint i;
+	gchar *input = NULL;
+	gchar *output = NULL;
+
+	if(!purple_account_get_bool(ma->account, "use-markdown", TRUE)) {

```

Comment:
Yes server does: I was just thinking that some (hardcore ;-)) users may not find this formatting that nice .. so having an option to disable it may be useful. But no - it is not necessary to have: whichever you prefer.